### PR TITLE
Feature/31 filter on a specific feature

### PR DESCRIPTION
**Pull request by GeReinhart**

### CODE REVIEW

**Comment by GeReinhart** [comment]
I don't understand why the Travis build is not working properly.
`sbt test` work fine on my machine

File: app/controllers/DocumentationController.scala [hunk 1]
```diff
@@ -68,16 +68,22 @@ class DocumentationController @Inject()(documentationRepository: DocumentationRe
...
+        val projectId = params(1)
         val project = projectRepository.findById(projectId)
         val projectName = project.map(_.name).getOrElse("")
+        val branchName = params.lift(2) match {
+          case Some(branchParam) if (! branchParam.isEmpty)  => branchParam
+          case _ => project.map(_.stableBranch).getOrElse("master")
+        }
 
-        val branchName = projectParam.split(">").lift(2).getOrElse(project.map(_.stableBranch).getOrElse("master"))
+        val mayFeatureFilter = params.lift(3)
```
Comment (notflorian) [inline app/controllers/DocumentationController.scala]: I'm not a fan this variable name. Maybe just rename it to 
```suggestion
        val featureFilter = params.lift(3)
```

File: app/controllers/DocumentationController.scala [hunk 2]
```diff
@@ -68,16 +68,22 @@ class DocumentationController @Inject()(documentationRepository: DocumentationRe
...
+        val mayTagsFilter = params.lift(4).map(tagsAsString => tagsAsString.split(",").toSeq  )
 
         hierarchy.lastOption.flatMap { hierarchyNode =>
           criteriaMap.get(hierarchyNode.id).flatten.map { criteria =>
-              buildDocumentation(criteria.hierarchy, Seq(documentationRepository.buildProjectDocumentation(ProjectCriteria(projectId, projectName, branchName))))
+              buildDocumentation(criteria.hierarchy, Seq(documentationRepository.buildProjectDocumentation(ProjectCriteria(projectId, projectName, branchName,mayFeatureFilter,mayTagsFilter))))
```
Comment (notflorian) [inline app/controllers/DocumentationController.scala]: Format issue
```suggestion
               buildDocumentation(criteria.hierarchy, Seq(documentationRepository.buildProjectDocumentation(ProjectCriteria(projectId, projectName, branchName, mayFeatureFilter, mayTagsFilter))))
```

File: app/filters/Filters.scala [hunk 3]
```diff
@@ -1,14 +1,15 @@
 package filters
 
 import akka.stream.Materializer
+import com.kelkoo.play.filters.AccessLogFilter
```
Comment (notflorian) [inline app/filters/Filters.scala]: Revert this change, as the accessLogFilter is Kelkoo specific and not opensourced

File: app/repository/FeatureRepository.scala [hunk 4]
```diff
@@ -82,9 +94,10 @@ class FeatureRepository @Inject()(db: Database, tagRepository: TagRepository, sc
   }
 
   def findAllByBranchId(branchId: Long): Seq[Feature] = {
-    db.withConnection { implicit connection =>
+    val s =  db.withConnection { implicit connection =>
```
Comment (notflorian) [inline app/repository/FeatureRepository.scala]: Revert this change as this local variable can be inlined
```suggestion
      db.withConnection { implicit connection =>
```

File: app/services/CriteriaService.scala [hunk 5]
```diff
@@ -1,27 +1,56 @@
...
+class CriteriaService @Inject()(hierarchyRepository: HierarchyRepository, projectRepository: ProjectRepository, branchRepository: BranchRepository, featureRepository: FeatureRepository, cache: SyncCacheApi) {
 
   def refreshCache(): Unit = Try(getCriteriasTree(true))
 
   def getCriterias(refresh: Boolean = false): Seq[Criteria] = {
     if (refresh) cache.remove(criteriasListCacheKey)
 
     cache.getOrElseUpdate(criteriasListCacheKey) {
+
+      val mapProjectIdProject = projectRepository.findAll().foldLeft(Map[String, Project]()) { (mapProjectIdProject, project) =>
```
Comment (notflorian) [inline app/services/CriteriaService.scala]: Simplify with
```suggestion
      val mapProjectIdProject = projectRepository.findAll().map(p => p.id -> p).toMap
```

File: app/services/CriteriaService.scala [hunk 6]
```diff
@@ -1,27 +1,56 @@
...
+        mapProjectIdProject + (project.id -> project)
+      }
+
+      val branches = branchRepository.findAll()
+      val mapProjectIdBranches = branches.groupBy(r => r.projectId).foldLeft(Map[String, Seq[Branch]]()) { (mapProjectIdBranches, branch) =>
```
Comment (notflorian) [inline app/services/CriteriaService.scala]: Simplify with
```suggestion
      val mapProjectIdBranches = branches.groupBy(_.projectId).map(b => b._1 -> b._.2.flatten).toMap
```

File: build.sbt [hunk 7]
```diff
@@ -1,9 +1,13 @@
 name := "the_gardener"
 
+val pom = xml.XML.load("pom.xml")
```
Comment (notflorian) [inline build.sbt]: Revert, as the maven packaging is Kelkoo specific

File: build.sbt [hunk 8]
```diff
@@ -1,9 +1,13 @@
...
+
+version := { pom \  "version" text}
+
 val jdkVersion = "1.8"
 scalaVersion := "2.12.7"
 
-lazy val root = (project in file(".")).enablePlugins(PlayScala)
+lazy val root = (project in file(".")).enablePlugins(PlayScala).enablePlugins(SbtWeb)
```
Comment (notflorian) [inline build.sbt]: Revert, as the SbtWeb plugin is not opensourced

File: build.sbt [hunk 9]
```diff
@@ -60,3 +67,7 @@ scalacOptions in Scapegoat += "-P:scapegoat:overrideLevels:TraversableHead=Warni
 evictionWarningOptions in update := EvictionWarningOptions.empty
 
 routesGenerator := InjectedRoutesGenerator
+
+JsEngineKeys.npmNodeModules in Assets := Nil
```
Comment (notflorian) [inline build.sbt]: Revert, as the SbtWeb plugin is not opensourced

File: build.sbt [hunk 10]
```diff
@@ -34,6 +38,7 @@ libraryDependencies ++= Seq(
   evolutions,
   jdbc,
   cacheApi,
+  "com.kelkoo.common" %% "playScalaCommon" % "2.1.2",
```
Comment (notflorian) [inline build.sbt]: Revert, as the lib playScalaCommon is not opensourced

File: conf/routes [hunk 11]
```diff
@@ -40,3 +40,6 @@ GET           /api/docs/*file                                 controllers.Assets
 
 
 
+#IsAlive
+GET     /HealthCheck/isAlive        com.kelkoo.play.controllers.IsAlive.isAlive(status: Option[String])
```
Comment (notflorian) [inline conf/routes]: Revert, as the lib playScalaCommon is not opensourced

File: app/services/CriteriaService.scala [hunk 12]
```diff
@@ -1,27 +1,56 @@
...
+        val branches = branch._2.foldLeft(Seq[Branch]())((branches, b) => branches :+ b)
+        mapProjectIdBranches + (branch._1 -> branches)
+      }
+
+      val mapBranchIdProjectFeaturePath = branches.foldLeft(Map[Long, String]()) { (mapBranchIdProjectFeaturePath, branch) =>
```
Comment (notflorian) [inline app/services/CriteriaService.scala]: Same here, simplify with .map

File: app/services/CriteriaService.scala [hunk 13]
```diff
@@ -1,27 +1,56 @@
...
+        mapBranchIdProjectFeaturePath + (branch.id -> mapProjectIdProject.get(branch.projectId).get.featuresRootPath )
+      }
+
+      val mapBranchIdFeaturePaths = featureRepository.findAllFeaturePaths().groupBy(r => r.branchId).foldLeft(Map[Long, Set[String]]()) { (mapBranchIdFeaturePaths, branchAndPath) =>
```
Comment (notflorian) [inline app/services/CriteriaService.scala]: Same here, simplify with .map

File: app/services/CriteriaService.scala [hunk 14]
```diff
@@ -1,27 +1,56 @@
...
+        val paths = branchAndPath._2.foldLeft(Set[String]()){(paths, p) =>
```
Comment (notflorian) [inline app/services/CriteriaService.scala]: Same here, simplify with .map

File: app/services/CriteriaService.scala [hunk 15]
```diff
@@ -1,27 +1,56 @@
...
+          val projectFeaturePath = mapBranchIdProjectFeaturePath.get(branchAndPath._1).get
+          val relativePath = p.path.substring(p.path.indexOf(projectFeaturePath)+projectFeaturePath.size+1)
```
Comment (notflorian) [inline app/services/CriteriaService.scala]: Global format issue in this file

File: frontend/src/app/_services/criteriasSelection.ts [hunk 16]
```diff
@@ -11,57 +11,151 @@ export class TupleHierarchyNodeSelector {
...
+      if ( split.length == 1 ){
+        this.features.push( new FeatureSelector( split.length, currentPath, currentPath )  ) ;
+      }else{
+        if ( split.length == 2 ){
+          var currentDirectory = split[0] ;
+          if (currentDirectory != lastDirectory){
+            this.features.push( new FeatureSelector( split.length - 1, currentDirectory, currentDirectory  +  ' :' )  ) ;
+            lastDirectory = currentDirectory  ;
+          }
+          this.features.push( new FeatureSelector( split.length, currentPath, split[1] )  ) ;
```
Comment (notflorian) [inline frontend/src/app/_services/criteriasSelection.ts]: Global formatting issue in this file

File: pom.xml [hunk 17]
```diff
@@ -0,0 +1,168 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (notflorian) [inline pom.xml]: Revert, as the maven packaging is Kelkoo specific

File: src/conf/application-prod.conf.erb [hunk 18]
```diff
@@ -0,0 +1,53 @@
+<% require 'kelkoo/erb_utils' %># This is the main configuration file for the application.
```
Comment (notflorian) [inline src/conf/application-prod.conf.erb]: Revert, as the erb conf is Kelkoo specific

File: test/features/generate_documentation/show_a_feature.feature [hunk 19]
```diff
@@ -18,7 +19,7 @@ Feature: Generate documentation
       | projectId     | hierarchyId |
       | suggestionsWS | .01.        |
 
-  @level_2_technical_details @nominal_case @valid
+  @level_2_technical_details @nominal_case @valid @ongoing
```
Comment (notflorian) [inline test/features/generate_documentation/show_a_feature.feature]: @ongoing tag

File: src/conf/logback.xml.erb [hunk 20]
```diff
@@ -0,0 +1,47 @@
+<% require 'kelkoo/erb_utils' %><configuration>
```
Comment (notflorian) [inline src/conf/logback.xml.erb]: Revert, as the erb conf is Kelkoo specific

File: src/conf/setenv.conf.erb [hunk 21]
```diff
@@ -0,0 +1,19 @@
+<% require 'kelkoo/erb_utils' %>#!/bin/sh
```
Comment (notflorian) [inline src/conf/setenv.conf.erb]: Revert, as the erb conf is Kelkoo specific

File: src/conf/default_settings.properties [hunk 22]
```diff
@@ -0,0 +1,63 @@
+#----------------------------------------------------------------------------------------------------------------
```
Comment (notflorian) [inline src/conf/default_settings.properties]: Revert, as the erb conf is Kelkoo specific

File: src/conf/application-prod.conf.erb (hunk 18 above)
Comment (notflorian) [inline src/conf/application-prod.conf.erb]: Revert, as the erb conf is Kelkoo specific

File: local-conf/application.dev.conf [hunk 23]
```diff
@@ -2,6 +2,6 @@ include "application.conf"
 
 db.default.driver=com.mysql.jdbc.Driver
```
Comment (notflorian) [inline local-conf/application.dev.conf]: Remove this file, as it shouldn't be commited on github

File: pom.xml (hunk 17 above)
Comment (notflorian) [inline pom.xml]: Revert, as the maven packaging is Kelkoo specific

File: src/conf/crontab [hunk 24]
```diff
@@ -0,0 +1 @@
+#crontab
```
Comment (notflorian) [inline src/conf/crontab]: Revert, as the crontab is Kelkoo specific

**Comment by notflorian** [comment]
> I don't understand why the Travis build is not working properly.
> `sbt test` work fine on my machine

Looks like a Tslint issue. Can you try in local `./node_modules/tslint/bin/tslint -t json -o report.json --project .` ?

**Comment by notflorian** [comment]
In fact, can you please remove from this pull request all the commit of the gitlab qa branch? (All those before the 27 Nov)
